internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,328 @@
+	log := log.FromContext(ctx)
+	log.Info("Reconciling CAPI cluster")
+
+	capiCluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, capiCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	patchBase := client.MergeFromWithOptions(capiCluster.DeepCopy(), client.MergeFromWithOptimisticLock{})
+
+	log = log.WithValues("cluster", capiCluster.Name)
+
+	// Wait for controlplane to be ready. This should never be false as the predicates
+	// do the filtering.
+	if !capiCluster.Status.ControlPlaneReady && !conditions.IsTrue(capiCluster, clusterv1.ControlPlaneReadyCondition) {
+		log.Info("clusters control plane is not ready, requeue")
+		return ctrl.Result{RequeueAfter: defaultRequeueDuration}, nil
+	}
+
+	// Collect errors as an aggregate to return together after all patches have been performed.
+	var errs []error
+
+	result, err := r.reconcile(ctx, capiCluster)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error reconciling cluster: %w", err))
+	}
+
+	if err := r.Client.Patch(ctx, capiCluster, patchBase); err != nil {
+		errs = append(errs, fmt.Errorf("failed to patch cluster: %w", err))
+	}
+
+	if len(errs) > 0 {
+		return ctrl.Result{}, errorutils.NewAggregate(errs)
+	}
+
+	return result, nil
+}
+
+func (r *CAPIImportManagementV3Reconciler) reconcile(ctx context.Context, capiCluster *clusterv1.Cluster) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// placeholder name/namespace must be set to be able to test this but it won't be used.
+	rancherCluster := &managementv3.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "c-",
+			Namespace: capiCluster.Namespace,
+			Labels: map[string]string{
+				capiClusterOwner:          capiCluster.Name,
+				capiClusterOwnerNamespace: capiCluster.Namespace,
+			},
+		},
+	}
+
+	rancherClusterList := &managementv3.ClusterList{}
+	selectors := []client.ListOption{
+		client.MatchingLabels{
+			capiClusterOwner:          capiCluster.Name,
+			capiClusterOwnerNamespace: capiCluster.Namespace,
+			ownedLabelName:            "",
+		},
+	}
+	err := r.RancherClient.List(ctx, rancherClusterList, selectors...)
+
+	if client.IgnoreNotFound(err) != nil {
+		log.Error(err, fmt.Sprintf("Unable to fetch rancher cluster %s", client.ObjectKeyFromObject(rancherCluster)))
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if len(rancherClusterList.Items) != 0 {
+		if len(rancherClusterList.Items) > 1 {
+			log.Info("More than one rancher cluster found. Will default to using the first one.")
+		}
+
+		rancherCluster = &rancherClusterList.Items[0]
+	}
+
+	if !capiCluster.ObjectMeta.DeletionTimestamp.IsZero() {
+		err := r.deleteDependentRancherCluster(ctx, capiCluster)
+		if err != nil {
+			return ctrl.Result{Requeue: true}, fmt.Errorf("error deleting associated managementv3.Cluster resources: %w", err)
+		}
+	}
+
+	if !rancherCluster.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
I've seen this failing but, looking at other projects, it looks like this is the proper way to handle the resource's deletion. Why would it not set a `DeletionTimestamp`? This could break the "un-import" strategy of Rancher Turtles, since CAPI cluster wouldn't be annotated and it would try to reconcile the Rancher Cluster again.

Reply:
Potentially if there isn't a finalizer.

Reply:
That's what I initially thought, but the resource's yaml shows multiple finalizers. I just re-checked and validated this:
```
API Version:  management.cattle.io/v3      
Kind:         Cluster                  
Metadata:                                
  Creation Timestamp:  2024-02-07T18:50:28Z
  Finalizers:                     
    controller.cattle.io/cluster-agent-controller-cleanup
    controller.cattle.io/cluster-scoped-gc 
    controller.cattle.io/cluster-provisioner-controller                                                
    controller.cattle.io/mgmt-cluster-rbac-remove
    wrangler.cattle.io/mgmt-cluster-remove 
    ```

Reply:
Maybe its a difference between wrangler and controller-runtime

Reply:
I was also thinking, that perhaps something else is updating the it? Before the update caused by the delete?

Reply:
I'd double check this. Even if I set a new finalizer that I never delete from the resource, it gets deleted. I expected this to make it stay in deleting status, so there must something that is interfering with this behavior.

Reply:
Ultimately the issue was that, when configuring the watcher, it was not translating the rancher cluster into the capi cluster properly [here](https://github.com/rancher-sandbox/rancher-turtles/blob/dcacd4ac52a868d4b532bdd465c8ce506f32ee18/internal/controllers/import_controller_v3.go#L91), which caused the reconcile to not behave as expected.